ksqldb-engine/src/main/java/io/confluent/ksql/schema/ksql/inference/SchemaRegistryTopicSchemaSupplier.java

```diff
@@ -74,67 +74,73 @@ public SchemaResult getKeySchema(
 
   @Override
   public SchemaResult getValueSchema(
-      final String topicName,
+      final Optional<String> topicName,
       final Optional<Integer> schemaId,
       final FormatInfo expectedFormat,
       final SerdeFeatures serdeFeatures
   ) {
     return getSchema(topicName, schemaId, expectedFormat, serdeFeatures, false);
   }
 
-  private SchemaResult getSchema(
-      final String topicName,
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public SchemaResult getSchema(
+      final Optional<String> topicName,
       final Optional<Integer> schemaId,
       final FormatInfo expectedFormat,
       final SerdeFeatures serdeFeatures,
       final boolean isKey
   ) {
-    try {
-      final String subject = getSRSubject(topicName, isKey);
+    if (!topicName.isPresent() && !schemaId.isPresent()) {
+      throw new IllegalArgumentException("One of topicName and schemaId should be provided");
+    }
 
+    try {
       final int id;
+      final String subject;
       if (schemaId.isPresent()) {
+        subject = topicName.map(s -> getSRSubject(s, isKey)).orElse(null);
         id = schemaId.get();
       } else {
+        subject = getSRSubject(topicName.get(), isKey);
         id = srClient.getLatestSchemaMetadata(subject).getId();
       }
 
       final ParsedSchema schema = srClient.getSchemaBySubjectAndId(subject, id);
-      return fromParsedSchema(topicName, id, schema, expectedFormat, serdeFeatures, isKey,
-          schemaId.isPresent());
+      return fromParsedSchema(topicName, id, schemaId, schema, expectedFormat, serdeFeatures,
```

Comment:
`id` and `schemaId` are the same. Can this become into one parameter?

Reply:
Will take a look how to refactor

Reply:
Actually `schemaId` is optional and is used to indicate whether schemaId is provided by user to fetched from latest schema. Maybe I can refactor to call it `isIdProvided`